MissingValueException for `javaCompiler` property with new `JdkImageWorkaround`

After trying `2.5.4`, we see this issue

```
FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':app'.
> Could not create task ':app:compileInternalDebugJavaWithJavac'.
   > Cannot query the value of task ':app:compileInternalDebugJavaWithJavac' property 'javaCompiler' because it has no value available.

// skipping stuff for brevity

Caused by: org.gradle.api.internal.provider.MissingValueException: Cannot query the value of task ':app:compileInternalDebugJavaWithJavac' property 'javaCompiler' because it has no value available.
        at org.gradle.api.internal.provider.AbstractMinimalProvider.get(AbstractMinimalProvider.java:85)
        at org.gradle.api.provider.Provider$get.call(Unknown Source)
        at org.gradle.android.workarounds.JdkImageWorkaround$_replaceCommandLineProvider_closure7$_closure12.doCall(JdkImageWorkaround.groovy:99)

```

--- a/src/main/groovy/org/gradle/android/workarounds/JdkImageWorkaround.groovy
+++ b/src/main/groovy/org/gradle/android/workarounds/JdkImageWorkaround.groovy
@@ -44,7 +44,7 @@ class JdkImageWorkaround implements Workaround {
             variant.javaCompileProvider.configure { JavaCompile task ->
                 def jdkImageInput = getJdkImageInput(task)
                 if (jdkImageInput != null) {
-                    setupExtractedJdkImageInputTransform(project, task.getJavaCompiler().map { it.metadata.installationPath })
+                    setupExtractedJdkImageInputTransform(project, getJvmHome(task))
                     replaceCommandLineProvider(task, jdkImageInput)
                     applyRuntimeClasspathNormalization(task.project)
                 }
@@ -96,12 +96,29 @@ class JdkImageWorkaround implements Workaround {
         def extractedJdkImage = jdkConfiguration.incoming.artifactView {viewConfiguration ->
             viewConfiguration.attributes {
                 it.attribute(ArtifactAttributes.ARTIFACT_FORMAT, JDK_IMAGE_EXTRACTED)
-                it.attribute(jdkId, task.javaCompiler.get().metadata.jvmVersion + task.javaCompiler.get().metadata.vendor)
+                it.attribute(jdkId, getJvmVersion(task) + getJvmVendor(task))
             }
         }.artifacts.artifactFiles
         def extractedJdkImageProvider = new ExtractedJdkImageCommandLineProvider(extractedJdkImage, jdkImageInput.jdkImage)
         task.options.compilerArgumentProviders.remove(jdkImageInput)
         task.options.compilerArgumentProviders.add(extractedJdkImageProvider)
+    }
+
+    static String getJvmVersion(JavaCompile task) {
+        return task.javaCompiler.map { it.metadata.jvmVersion }
+            .orElse(SystemPropertiesCompat.getString("java.version", task.project, null))
+    }
+
+    static String getJvmVendor(JavaCompile task) {
+        return task.javaCompiler.map { it.metadata.vendor }
+            .orElse(SystemPropertiesCompat.getString("java.vendor", task.project, null))
+    }
+
+    static Provider<Directory> getJvmHome(JavaCompile task) {
+        def defaultJvmHome = task.project.objects.directoryProperty()
+        defaultJvmHome.set(new File(SystemPropertiesCompat.getString("java.home", task.project, null)))
+        return task.javaCompiler.map { it.metadata.installationPath }
+            .orElse(defaultJvmHome)
     }
 
     @Override

--- a/src/main/groovy/org/gradle/android/workarounds/SystemPropertiesCompat.groovy
+++ b/src/main/groovy/org/gradle/android/workarounds/SystemPropertiesCompat.groovy
@@ -3,6 +3,7 @@ package org.gradle.android.workarounds
 import groovy.transform.CompileStatic
 import org.gradle.android.Versions
 import org.gradle.api.Project
+import org.gradle.api.provider.Provider
 import org.gradle.util.GradleVersion
 
 @CompileStatic
@@ -31,14 +32,7 @@ class SystemPropertiesCompat {
      */
     static boolean getBoolean(String key, Project project, Boolean absentValue) {
         if (project != null) {
-            def systemProperty = project.providers.systemProperty(key)
-
-            if (Versions.gradle(project.gradle.gradleVersion) < GradleVersion.version("7.4")) {
-                systemProperty = systemProperty.forUseAtConfigurationTime()
-            }
-
-            systemProperty
-                .map{
+            getSystemProperty(key, project).map {
                     try {
                         return Boolean.parseBoolean(it)
                     } catch (IllegalArgumentException | NullPointerException ignored) {
@@ -49,5 +43,23 @@ class SystemPropertiesCompat {
         } else {
             return Boolean.getBoolean(key)
         }
+    }
+
+    static String getString(String key, Project project, String absentValue) {
+        if (project != null) {
+            getSystemProperty(key, project).getOrElse(absentValue)
+        } else {
+            return System.getProperty(key)
+        }
+    }
+
+    private static Provider<String> getSystemProperty(String key, Project project) {
+        def systemProperty = project.providers.systemProperty(key)
+
+        if (Versions.gradle(project.gradle.gradleVersion) < GradleVersion.version("7.4")) {
+            systemProperty = systemProperty.forUseAtConfigurationTime()
+        }
+
+        return systemProperty
     }
 }

--- a/src/test/groovy/org/gradle/android/JdkImageWorkaroundTest.groovy
+++ b/src/test/groovy/org/gradle/android/JdkImageWorkaroundTest.groovy
@@ -5,6 +5,7 @@ import org.gradle.api.JavaVersion
 import org.gradle.testkit.runner.BuildResult
 import org.gradle.testkit.runner.TaskOutcome
 import org.junit.Assume
+import spock.lang.Issue
 
 @MultiVersionTest
 class JdkImageWorkaroundTest extends AbstractTest {
@@ -201,5 +202,45 @@ class JdkImageWorkaroundTest extends AbstractTest {
 
         then:
         buildResult.output.contains("CommandLineArgumentProvider is ExtractedJdkImageCommandLineProvider")
+    }
+
+    @Issue("https://github.com/gradle/android-cache-fix-gradle-plugin/issues/307")
+    def "jdkImage is normalized when toolchain is not specified"() {
+        def androidVersion = TestVersions.latestAndroidVersionForCurrentJDK()
+        def gradleVersion = TestVersions.latestSupportedGradleVersionFor(androidVersion)
+        SimpleAndroidApp.builder(temporaryFolder.root, cacheDir)
+            .withAndroidVersion(androidVersion)
+            .withKotlinDisabled()
+            .withSourceCompatibility(JavaVersion.VERSION_1_9)
+            .build()
+            .writeProject()
+
+        when:
+        BuildResult buildResult = withGradleVersion(gradleVersion.version)
+            .withProjectDir(temporaryFolder.root)
+            .withArguments(
+                "clean", "assemble",
+                "--build-cache"
+            ).build()
+
+        then:
+        buildResult.task(':app:compileDebugJavaWithJavac').outcome == TaskOutcome.SUCCESS
+        buildResult.task(':app:compileReleaseJavaWithJavac').outcome == TaskOutcome.SUCCESS
+        buildResult.task(':library:compileDebugJavaWithJavac').outcome == TaskOutcome.SUCCESS
+        buildResult.task(':library:compileReleaseJavaWithJavac').outcome == TaskOutcome.SUCCESS
+
+        when:
+        buildResult = withGradleVersion(gradleVersion.version)
+            .withProjectDir(temporaryFolder.root)
+            .withArguments(
+                "clean", "assemble",
+                "--build-cache"
+            ).build()
+
+        then:
+        buildResult.task(':app:compileDebugJavaWithJavac').outcome == TaskOutcome.FROM_CACHE
+        buildResult.task(':app:compileReleaseJavaWithJavac').outcome == TaskOutcome.FROM_CACHE
+        buildResult.task(':library:compileDebugJavaWithJavac').outcome == TaskOutcome.FROM_CACHE
+        buildResult.task(':library:compileReleaseJavaWithJavac').outcome == TaskOutcome.FROM_CACHE
     }
 }

